### PR TITLE
Implement minimum stat and skill requirements for items

### PR DESCRIPTION
**Pull request by mugling**
This PR parses minimum required stats and/or skill levels from JSON and provides `player::can_use` to check item requirements are met. Initially it is used to deprecate `STR8_DRAW`, `STR10_DRAW` and `STR12_DRAW` however the implementation is generic enough that much more is possible.

Presently item requirements are checked only for ranged weapons and comments are sought as to what other circumstances in which they should apply. Possible candidates could include restricting wearing of items (too weak to wear an armour) or tool usage (need skill in fabrication before using an acetylene torch).

Also comments as to existing stats/skills checks that could use this functionality, especially any currently dependent upon flags. Possibly the `skill_required` field for gunmods could be converted to use the new code?

### CODE REVIEW

File: src/item.cpp [hunk 1]
```diff
@@ -3773,29 +3773,15 @@ int item::gun_range( const player *p ) const
...
-        }
-        if( p->str_cur < 12 ) {
-            ret -= 2 * ( 12 - p->str_cur );
-        }
+    if( !p->can_use( *this, false ) ) {
+        return 0;
     }
+
+    // Reduce bow range until player has twice minimm required strength
+    if( gun_skill() == skill_id("archery") ) {
```
Comment (Coolthulhu) [inline src/item.cpp]: That's hacky, should be controlled in some other way. For example, a flag.

File: src/player.h [hunk 2]
```diff
@@ -686,6 +686,11 @@ class player : public Character, public JsonSerializer, public JsonDeserializer
         /** Creates the UI and handles player input for picking martial arts styles */
         bool pick_style();
         /**
+         * Checks if player stats and skills meet minimum requirements for the item
+         * @param interactive controls whether informative messages are printed if item requirements not met
+         */
+	bool can_use( const item& it, bool interactive = true ) const;
```
Comment (Coolthulhu) [inline src/player.h]: Tabs

File: src/game.cpp [hunk 3]
```diff
@@ -10892,17 +10888,7 @@ void game::plfire( bool burst, const tripoint &default_target )
...
-                strength_needed = 8;
-            }
-            if (u.weapon.has_flag("STR10_DRAW")) {
-                strength_needed = 10;
-            }
-            if (u.weapon.has_flag("STR12_DRAW")) {
-                strength_needed = 12;
-            }
-            u.mod_stat("stamina", strength_needed * -2);
+            u.mod_stat( "stamina", std::min( u.weapon.type->min_str, 6 ) * -2 );
```
Comment (Coolthulhu) [inline src/game.cpp]: Why cap it?

**Comment by kevingranade** [comment]
Armor wearing should just be capped by weight limits.
Likewise melee and thrown weapons become unusable because of poor strength/weight ratio leading to slow attack speeds, but I feel like that should just result in a "it will take a long time to swing this weapon" message on wield if it passes some threshhold.
Bows really are a special case where you need a certain level level of strength to use them at all.

File: src/game.cpp (hunk 3 above)
Comment (mugling) [inline src/game.cpp]: For equivalence with the previous code?

**Comment by mugling** [comment]
Can I consider the refactors in `game::plfire` acceptable - I want to continue refactoring that function in a separate PR as part of the work towards implementing magazines?

**Comment by sparr** [comment]
I agree with kevin about the specific ideas put forward, but I think having a generic requirements capability could still be useful, and I'm all in favor of getting rid of flags with variables in their name.

**Comment by Night-Pryanik** [comment]
> need skill in fabrication before using an acetylene torch

I think mandatory full restriction is too rough. I can screw things up, I can hurt myself trying to use tool, I can destroy tool or thing I'm using torch on. But I should have a right to try! With penalties if my skill isn't high enough, but nevertheless.

**Comment by AppleConnoiseur** [comment]
Not entirely a fan of being outright disallowed from using equipment.
Would be better if it worked liked books at being a soft cap where it is harder to read a book if your intelligence doesn't match.

**Comment by SeanMirrsen** [comment]
I also don't think it's entirely correct to outright disallow the use of particular items. Even in the case of bows, all you'd have is a hilarious decrease in effectiveness (i.e. range). If the strength is so inadequate that the maximum range is zero, that's exactly what should happen then. With, likely, a message along the lines of "You are too weak to draw this bow!" appearing when you use the `f`ire command to let the player know what's going on. If the strength is merely too low to use the bow to its full capacity, the message should instead say that you can't draw the bowstring fully, or something along those lines.

**Comment by kevingranade** [comment]
If you can't draw a bow fully, you can't aim properly, the range would be
lowered, and the damage would be negligible. At a certain point it becomes
not worthwhile to clutter up the code in order to make a distinction
between, "can't do it" and "extremely ineffective".

**Comment by SeanMirrsen** [comment]
I wouldn't say it will add clutter. Scaling bow range and accuracy with strength is a mechanic that makes sense, so it will exist anyway. A system that alerts the player when they aren't getting the full use out of something they're doing (like when reading at low intelligence or drawing a bow with low strength), also makes sense to have. This would just combine the two, leading to emergent design rather than an arbitrary cutoff.

**Comment by kevingranade** [comment]
Er, more strength does not lead to more power out of a bow.  It stores and
releases a certain amount of energy based on its construction, and the
archer either has enough strength or not.

**Comment by SeanMirrsen** [comment]
More, no. Less, yes. A bow can be pulled back under its limit, exactly because all it does is store energy. You get far less range and power out of it, and it decreases exponentially as you get less and less of its maximum draw, but there isn't a hard cutoff between "can bend it fully" and "can't bend it at all".

**Comment by kevingranade** [comment]
After a little reading, it seems I was mistaken, there are at least some
sources indicating that under- or over-draw are used from time to time, so
the question is just one of significance.

**Comment by SeanMirrsen** [comment]
As far as significance, I'd have to agree that in the grand scheme of things it'd be marginal. Bows are just one weapon type, and not that frequently used outside of specific circumstances, and even then the scaling would often not have a chance to significantly come into play (i.e. if a character is too strong for a shortbow, chances are they're just right for a longbow or recurve bow or composite bow, etc).

However, that does not take into account the implications of a generalized system of skill/stat requirements. You can't, for example, be too dexterous for a sword, but you can be not dexterous enough for a set of nunchucks, leading to occasional self-hits, with chance increasing as you get below limit in dexterity. Just for example.

And really, given the scope of the game now, any one mechanic could seem "of marginal significance" in comparison. :)
----------------------
And speaking of overdraw, the current implementation apparently forbids it. It would be neat to be able to get a few extra tiles of range out of a bow, at the cost of potential damage to it.
Also a warning for the player when their bow range is being reduced because of their low strength could be nice.

And while I'm here making suggestions I spotted STR_RELOAD in the list of tags, and remembered that some crossbows - ones lacking a ratchet mechanism to draw the string - have to be reloaded by pulling on the string with your hands, and wouldn't be "slower to reload" so much as "impossible to reload" if you happen to have the body strength of a five-year-old. So that's about the only instance of a "hard stat limit" I can think of for now.

**Comment by AppleConnoiseur** [comment]
Overdrawing bows would be a interesting mechanic. Either use it for longer range, greater accuracy at normal range and greater armor penetration when in a 5-7 tile distance. Could maybe have a chance to damage the bow if it is a flimsy construction (i.e self bow) unless specifically constructed to allow that.
Could be expanded to throwing weapons too i guess.

**Comment by mugling** [comment]
> but I think having a generic requirements capability could still be useful, and I'm all in favor of getting rid of flags with variables in their name.

This summaries well this PR's goal as deprecating some flags containing numeric values. Further uses would be a welcome bonus.

> At a certain point it becomes not worthwhile to clutter up the code in order to make a distinction between, "can't do it" and "extremely ineffective"

I would agree. That certain point is where we specify the **minimum** requirements. This approach is simple to implement and easiest to adjust when deciding on balance.

> It should be an emergent thing.

Nothing in this PR precludes emergent behavior. Indeed with `STR_DRAW` range improves proportional to strength providing the minimum requirements are met.

> I can hurt myself trying to use tool, I can destroy tool or thing I'm using torch on. But I should have a right to try! With penalties if my skill isn't high enough, but nevertheless.

Balance could be problematic without well chosen penalties. Simple task failure encourages grinding and we all agree tedium is not an appropriate balancing factor. Alternatively if the penalty is too harsh (eg IRL incorrect use of an acetylene torch can result in literal permadeath) there is no incentive for the player to try at low skill levels creating the defacto minimum @kevingranade describes.

Having minimum requirements and also scaling dependent upon stats/skills are not mutually exclusive ideas. Unless there is a clear way to relate them this PR only intends however to tackle the former.

**Pull request merged**